contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,773 @@
+    Eg. coverAssetsFallback = 3 (in binary 11) means assets at index 0 and 1 are supported.
+  */
+  uint32 public coverAssetsFallback;
+
+  // Global active cover amount per asset.
+  mapping(uint24 => uint) public override totalActiveCoverInAsset;
+
+  bool public coverAmountTrackingEnabled;
+  bool public activeCoverAmountCommitted;
+
+
+  /* ========== CONSTRUCTOR ========== */
+
+  constructor(
+    IQuotationData _quotationData,
+    IProductsV1 _productsV1,
+    address _coverNFT,
+    address _stakingPoolImplementation
+  ) {
+
+    // initialize immutable fields only
+    quotationData = _quotationData;
+    productsV1 = _productsV1;
+    coverNFT = _coverNFT;
+
+    stakingPoolProxyCodeHash = CoverUtilsLib.calculateProxyCodeHash();
+    stakingPoolImplementation = _stakingPoolImplementation;
+  }
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+
+  /// @dev Migrates covers from V1. Meant to be used by Claims.sol and Gateway.sol to allow the
+  /// users of distributor contracts to migrate their NFTs.
+  ///
+  /// @param coverId     V1 cover identifier
+  /// @param fromOwner   The address from where this function is called that needs to match the
+  /// @param toNewOwner  The address for which the V2 cover NFT is minted
+  function migrateCoverFromOwner(
+    uint coverId,
+    address fromOwner,
+    address toNewOwner
+  ) external override onlyInternal {
+    _migrateCoverFromOwner(coverId, fromOwner, toNewOwner);
+  }
+
+  /// @dev Migrates covers from V1
+  ///
+  /// @param coverId     V1 cover identifier
+  /// @param fromOwner   The address from where this function is called that needs to match the
+  /// @param toNewOwner  The address for which the V2 cover NFT is minted
+  function _migrateCoverFromOwner(
+    uint coverId,
+    address fromOwner,
+    address toNewOwner
+  ) internal {
+
+    CoverUtilsLib.migrateCoverFromOwner(
+      CoverUtilsLib.MigrateParams(
+        coverId,
+        fromOwner,
+        toNewOwner,
+        ICoverNFT(coverNFT),
+        quotationData,
+        tokenController(),
+        productsV1
+      ),
+      _products,
+      _productTypes,
+      _coverData,
+      _coverSegments
+    );
+  }
+
+  /// @dev Migrates covers from V1. Meant to be used by EOA Nexus Mutual members
+  ///
+  /// @param coverIds    Legacy (V1) cover identifiers
+  /// @param toNewOwner  The address for which the V2 cover NFT is minted
+  function migrateCovers(uint[] calldata coverIds, address toNewOwner) external override {
+    for (uint i = 0; i < coverIds.length; i++) {
+      _migrateCoverFromOwner(coverIds[i], msg.sender, toNewOwner);
+    }
+  }
+
+  function buyCover(
+    BuyCoverParams memory params,
+    PoolAllocationRequest[] memory allocationRequests
+  ) external payable override onlyMember whenNotPaused returns (uint /*coverId*/) {
+
+    require(_products.length > params.productId, "Cover: Product not found");
+    Product memory product = _products[params.productId];
```

Comment:
The same read is performed in `allocateCapacity` which is probably going to get inlined. Try passing the `product` as a parameter to `_buyCover` below to avoid an extra storage read.

Reply:
i'll pass on this optimization for now, seems small, also in editCover it's only read once